### DISCUSSION TITLE
Welcome to Box2D 3.0 Discussions!

### DISCUSSION THREAD

**Discussion by erincatto** [erincatto/box2c#13] [Announcements]
## 👋 Welcome to Box2D 3.0!
I've setup discussions to get feedback on the development of Box2D version 3.0. Feel free to give feedback on any aspect of 3.0. In particular I would like feedback on:
  * build system using cmake and vcpkg
  * C API
  * C usage
  * folder and file layout
  * features ideas to make Box2D better for everyone

**Comment by GloriousPtr** [comment]
I would love to get the Macros for profiling.

**Comment by 1vanK** [comment]
It would be cool to have capsule shape (for a character controller). Of course, the capsule can be made up of a rectangle and two circles, however, but in this case, random collisions of the corners of the rectangle and the polygons of a level often occur.

**Comment by erincatto** [comment]
Capsule shapes are already in v3.0. I agree they are useful for characters.

**Reply by 1vanK** [reply]
https://github.com/erincatto/box2c/blob/main/src/manifold.c#L83

By the way, I noticed that b2Transform is passed by value, and not by a constant pointer. Is it intentional?

**Reply by erincatto** [reply]
Yes. I plan to pass anything 16 bytes or less by value.

**Reply by 1vanK** [reply]
Interesting, if there are any articles about this?

**Reply by erincatto** [reply]
This sort of thing you can investigate on godbolt. Check if the values are passed in register or require a memory lookup. Registers are faster.

**Comment by GloriousPtr** [comment]
Adding buoyancy support out of the box would be great.

**Reply by ice1000** [reply]
Does that imply liquid support?

**Reply by masonova1** [reply]
This could mean just a static "liquid" region or object group, but then, that seems like a job for contact handlers? Unless you meant something to handle that boiler-plating.

**Comment by Ok23** [comment]
Create inactive orphan fixtures, attach and detach it from body without destroying it (for preserve pointers to it and userdata) or just migrate fixtures from body to body without destroying.
Change shape type in existing fixture. 
For example it helps dynamic entities attach to each other in game and to change their forms dynamically. And maybe merge bodies ?

**Comment by nagolove** [comment]
Why you are using C instead of C++? I think choice based on bindings necessity to python, luajit etc.

**Reply by erincatto** [reply]
I write about the decision to use C here: https://box2d.org/posts/2023/01/starting-box2d-3.0/

**Reply by nagolove** [reply]
Thanks for reply! Now I am use mostly pure C in my projects. For physics I use Chipmunk2D which based on earlier version of Box2xd. It has problems with tunelling and it has not `bullet` flag for bodies. I think box2c will be a great project. Also I interested in multithreaded decision for objects-fulled scenes.

**Comment by GrauBlitz** [comment]
I love Box2D since years and I am highly interested in a pure C library (C > C++ >:-) ) is their anyway to support you outside of Github Sponsorship (I don't want to give Github my full data...)?

Thank you for all your work, I'm really hyped!

**Reply by erincatto** [reply]
I set this up. Not sure if it works. https://www.patreon.com/Box2D

Thank you!

**Reply by GrauBlitz** [reply]
I can look at it, but I think you need to create member levels or enable custom amount pledges. I don't see any option to send some money currently.

**Comment by bluesillybeard** [comment]
Small question, I am making a project with some requirements:
- the majority (like 99%) of static objects will be small squares since the game is a tile grid (similar to Terarria or Starbound)
    * I plan on using an algorithm to combine tile hitboxes into fewer larger ones
- fabric (and maybe hair) "simulations" so I can make the game feel fluid without billions of animations
- Procedural animation as well but that can be more or less independent of the physics system I use
- C api since I am using the Zig programming language which does not play nicely with C++

Is box2D v3 in a position where I can use it for those things, or should I create a C wrapper for v2 and use that?

**Reply by erincatto** [reply]
I plan for v3 to be ready for testing by the end of 2023 and releasing first quarter of 2024.

**Reply by bluesillybeard** [reply]
In that case I'll work on other aspects of the project for now, thanks a ton!

**Comment by ShockSlayer** [comment]
I think something like a "b2_math_extra.h" would be extremely helpful for a lot of people. The idea would be to provide some extra utility functions that aren't necessarily used in the Box2D library itself, but can be used in conjunction with it, using the common data formats like b2Vec2, b2Shape, etc.

An example might be something akin to a function for finding the area of overlap between two shapes. There are probably other similar concepts that would apply here.

**Reply by bluesillybeard** [reply]
Agreed! I don't think it would be necessary, (especially since I usually end up having to make my own math utilities anyway), but it would certainly help a lot. Even just exposing some of the internal math functions would go a long way.

**Comment by simonvanbernem** [comment]
In my game, I'm using dynamically generated geometry and a couple of times, this has lead to box2d crashing because a particular piece of geometry managed to trigger an assert inside box2d. Things like polygon vertices being welded until none are left are handled gracefully now by reporting an error to the user code, to which it can react. This is good.

Today I found a crash in ResetMassData (see https://github.com/erincatto/box2d/issues/756) where an assert triggers for valid inputs probably due to numerical precision issues.

For v3, it would be nice if these kinds of asserts couldn't happen. I think an assert for a misuse of the API is perfectly reasonable, but if the API accepted the inputs, it shouldn't crash on them. Maybe there is a way to handle this more systematically? Not sure. Maybe it's just a case of "report all of them until there's enough safety margin everywhere", although I'm not sure that this approach works reliably with floating point stuff. Or maybe one could have fallback values that at least keep the simulation running in such cases.

**Reply by erincatto** [reply]
v3 provides hull validation using `b2ValidateHull`, so you can avoid the assert on bad data. If the hull is valid yet mass computation still fails then that is a bug I need to fix.

I think it is important that Box2D asserts under failure, otherwise things will never get fixed. However, I should provide tools for the user to verify their data. You will be able to use a build mode that disables assertions if you like, but this does not guarantee the program will not crash or behave badly.

I'm adding more testing facilities to v3 and I expect to have pretty good coverage upon release and increasing over time.
